chips/rp2040/src/pio.rs

```diff
@@ -0,0 +1,1233 @@
+        self.registers.sm[sm_number as usize].execctrl.set(execctrl);
+        self.registers.sm[sm_number as usize].pinctrl.set(pinctrl);
+    }
+
+    /// Immediately execute an instruction on a state machine.
+    pub fn sm_exec(&self, sm_number: SMNumber, instr: u32) {
+        self.registers.sm[sm_number as usize]
+            .instr
+            .modify(SMx_INSTR::INSTR.val(instr));
+    }
+
+    /// Write a word of data to a state machine’s TX FIFO.
+    pub fn sm_put(&self, sm_number: SMNumber, data: u32) {
+        self.registers.txf[sm_number as usize].set(data);
+    }
+
+    /// Wait until a state machine's TX FIFO is empty, then write a word of data to it.
+    pub fn sm_put_blocking(&self, sm_number: SMNumber, data: u32) {
+        while self.registers.fstat.read(FSTAT::TXFULL0) != 0 {}
+        self.registers.txf[sm_number as usize].set(data);
+    }
+
+    /// Restart a state machine.
+    pub fn restart_sm(&self, sm_number: SMNumber) {
+        // SET Reg
+        match sm_number {
+            SMNumber::SM0 => self.set_registers.ctrl.modify(CTRL::SM0_RESTART::SET),
+            SMNumber::SM1 => self.set_registers.ctrl.modify(CTRL::SM1_RESTART::SET),
+            SMNumber::SM2 => self.set_registers.ctrl.modify(CTRL::SM2_RESTART::SET),
+            SMNumber::SM3 => self.set_registers.ctrl.modify(CTRL::SM3_RESTART::SET),
+        }
+    }
+
+    /// Clear a state machine’s TX and RX FIFOs.
+    pub fn sm_clear_fifos(&self, sm_number: SMNumber) {
+        // XOR Reg
+        self.xor_registers.sm[sm_number as usize]
+            .shiftctrl
+            .modify(SMx_SHIFTCTRL::FJOIN_RX::SET);
+        // XOR Reg
+        self.xor_registers.sm[sm_number as usize]
+            .shiftctrl
+            .modify(SMx_SHIFTCTRL::FJOIN_RX::SET);
+    }
+
+    /// Restart a state machine's clock divider.
+    pub fn sm_clkdiv_restart(&self, sm_number: SMNumber) {
+        match sm_number {
+            // SET Reg
+            SMNumber::SM0 => self.set_registers.ctrl.modify(CTRL::CLKDIV0_RESTART::SET),
+            SMNumber::SM1 => self.set_registers.ctrl.modify(CTRL::CLKDIV1_RESTART::SET),
+            SMNumber::SM2 => self.set_registers.ctrl.modify(CTRL::CLKDIV2_RESTART::SET),
+            SMNumber::SM3 => self.set_registers.ctrl.modify(CTRL::CLKDIV3_RESTART::SET),
+        }
+    }
+
+    /// Adds a program to PIO.
+    /// Call this with add_program(include_bytes!("path_to_file")).
+    pub fn add_program(&self, program: &[u8]) {
+        self.clear_instr_registers();
+        let iter = program.chunks(2);
+        let mut x = 0;
+        for i in iter {
+            self.registers.instr_mem[x]
+                .instr_mem
+                .modify(INSTR_MEMx::INSTR_MEM.val((i[0] as u32) << 8 | (i[1] as u32)));
+            x += 1;
+            if x == 32 {
+                break;
+            }
+        }
+    }
+
+    /// Clears all of a PIO instance's instruction memory.
+    pub fn clear_instr_registers(&self) {
+        for i in 0..31 {
```

Comment:
[Rust ranges are only half open](https://doc.rust-lang.org/std/ops/struct.Range.html), i.e. they exclude the second value from the iteration. I suspect this should be `0..=31` or `0..32`. Even better would be defining a constant representing the number of instructions that fit in memory per PIO instance.

Reply:
Thank you, switched it to the constant used for the registers, NUMBER_INSTR_MEMORY_LOCATIONS - also used it in add_program to make sure only 32 instructions are uploaded per program.